From d19bff928d379afbc417cd317b9e702598bce285 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 01:52:10 -0400
Subject: [iommu] amd: Take mmap_sem when calling get_user_pages

Message-id: <20140905015210.31297.64966.stgit@gir.stowe>
Patchwork-id: 91357
O-Subject: [RHEL7.1 PATCH 64/65] iommu/amd: Take mmap_sem when calling get_user_pages
Bugzilla: 1129880 1087643
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1087643
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1129880
Upstream Status: 4378d99295fae5fd42b0ed1a3eeef2a1684e0c17

commit 4378d99295fae5fd42b0ed1a3eeef2a1684e0c17
Author: Jay Cornwall <jay.cornwall@amd.com>
Date:   Mon Apr 28 17:27:46 2014 -0500

    iommu/amd: Take mmap_sem when calling get_user_pages

    get_user_pages requires caller to hold a read lock on mmap_sem.

    Signed-off-by: Jay Cornwall <jay.cornwall@amd.com>
    Signed-off-by: Suravee Suthikulpanit <suravee.suthikulpanit@amd.com>
    Signed-off-by: Joerg Roedel <joro@8bytes.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/amd_iommu_v2.c b/drivers/iommu/amd_iommu_v2.c
index 5208828..203b2e6 100644
--- a/drivers/iommu/amd_iommu_v2.c
+++ b/drivers/iommu/amd_iommu_v2.c
@@ -504,8 +504,10 @@ static void do_fault(struct work_struct *work)
 
  write = !!(fault->flags & PPR_FAULT_WRITE);
 
+ down_read(&fault->state->mm->mmap_sem);
  npages = get_user_pages(fault->state->task, fault->state->mm,
     fault->address, 1, write, 0, &page, NULL);
+ up_read(&fault->state->mm->mmap_sem);
 
  if (npages == 1) {
   put_page(page);
-- 
1.7.1